### ISSUE TITLE
ssoSilent in IE11 results in "BrowserAuthError: monitor_window_timeout: Token acquisition in popup failed due to timeout."

### ISSUE THREAD

**Issue by daniel-seitz** [AzureAD/microsoft-authentication-library-for-js#2225]
>> Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [x] `@azure/msal-browser@2.1.0` (& @azure/msal-common@1.2.0)
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

**Important**: Please fill in your exact version number above, e.g. `msal@1.1.3`.

## Framework
Angular 10

## Description
a. Upon loading (refreshing) the app my code initializes the auth via new PublicClientApplication and then "handleRedirectPromise"
b. Additionally, to determine and set the initial logged-in state, I use the ssoSilent call.

So when clicking a login button, I expect to be redirected to MS and after successful login, a. should handle the result.
When refreshing the page, b. should determine the current state and set it.

Tested in Chrome and Edge - works.
In IE11 both options lead to the same error.
In a. I can't really see if it comes from another call than ssoSilent since IE always closes my console and doesnt show a history.

Should this library (and ssoSilent) work in IE11?

## Error Message
First: 
SEC7131: Security of a sandboxed iframe is potentially compromised by allowing script and same origin access.

After a few seconds: 
"BrowserAuthError: monitor_window_timeout: Token acquisition in popup failed due to timeout."

-- I didnt configure "Popup" anywhere.
-- The solution given in https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/2213 does NOT resolve this issue.

## Security
- [ ] Is this issue security related?

## Regression
- [ ] Did this behavior work before? 
Version:

## MSAL Configuration

            auth: {
                clientId: ...,
                authority: ...,
                knownAuthorities: ...,
                redirectUri: ...,
                validateAuthority: true,
                navigateToLoginRequestUrl: false,
            },
            cache: {
                cacheLocation: 'localStorage',
                storeAuthStateInCookie: false,
            },
            system: {
                logger: new Logger({
                    ...
                }),
                tokenRenewalOffsetSeconds: 300,
                navigateFrameWait: 500,
            },

## Expected behavior
No error given, ssoSilent resolve into AuthenticationResult

## Browsers/Environment
- [ ] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari
- [x] IE11
- [ ] Other (Please add browser name here)

**Comment by jmckennon** [comment]
> Should this library (and ssoSilent) work in IE11?

Yes.

> After a few seconds:
"BrowserAuthError: monitor_window_timeout: Token acquisition in popup failed due to timeout."

This is a known issue that the error message is incorrect, it should say "iframe" instead of "popup".

I'm going to be trying to reproducing this today/Monday and will report back, this could definitely be a bug.

**Comment by daniel-seitz** [comment]
@jmckennon Thank you.

In my 'determineInitialState'-method I check if this is being called in an iframe and return. Otherwise I continue.
Looks like ssoSilent then creates the iframe and then times out.

Let me know if you need further info.

Daniel

**Comment by daniel-seitz** [comment]
Just to note if helpful: When I call ssoSilent again and again, the pattern seems to be:

1. doesnt work
2. works
3. doesnt work
4. works
5. doesnt work
6. works
7. doesnt work
8. works
...

**Comment by jmckennon** [comment]
@daniel-seitz Sorry for the delay, I haven't been able to reproduce these issues you're seeing. Can you provide both the `msal` logs you're seeing and also the rest of your `msal` code usage?

**Comment by daniel-seitz** [comment]
Hi @jmckennon, no worries! We all have things to do..

I think that it is actually another issue.

When I load the page with different situations - since we only have loginRedirect, these are: 
- page refresh (when logged out),
- load (after login),
- load (after logout and then click back without choosing an account to sign out of),
- load (after logout and chose an account)
- _page refresh (when logged in)_

**The last one doesn't work correctly.** it says not-authenticated

I tried to circumvent it with the code below.

However this doesn't really work correct either (I just found out yesterday. Assumption: since there are actually two competing processes and most of the time the correct one wins). I guess the error mentioned in this ticket might be a consequential error - _so this one should be my bad._

**The gist here:**

    const msalConfig = ... // see above

    const resultHandler = (result: AuthenticationResult) => {
        if (!result) {
            // **if I do set isAuth$ to false here then refresh when logged in does not work!!**
            return;
        }
        this.isAuth$.next(true);
    };


    this.authApp = new PublicClientApplication(msalConfig);

    this.authApp
        .handleRedirectPromise()
        .then(resultHandler);


    // **everything below might be unnecessary:**
    // the intention of the code below is to figure out if we are logged in on page refresh
    if (isInIframe()) {
        return;
    }

    let initialState = false;

    const accounts = this.authApp.getAllAccounts();

    if (accounts.length === 1) {
        const ssoRequest = {
            scopes: [],
            loginHint: accounts[0].username,
        } as AuthorizationUrlRequest;

        try {
            const authResult = await this.authApp.ssoSilent(ssoRequest);
            initialState = !!authResult;
        } catch (error) {
            // too many requests, network error or so
        }
    }

    this.isAuth$.next(initialState);

@jmckennon can you confirm:
1.  the page refresh when logged in is a bug?
2. I saw that the retrieved access tokens have the ms graph ressource uri as aud claim and not the applications clientId, is this also a bug?

**Comment by daniel-seitz** [comment]
I now circumvent the bug in the comment above with this code. Unfortunately the bug in IE11 still occurs:

    const msalConfig = ... // see bug description

    const resultHandler = (result: AuthenticationResult) => {
        if (!result) {
            checkAndSetInitialState();  // PREVENT REFRESH WHEN LOGGED IN BUG
            return;
        }
        this.isAuth$.next(true);
    };

    this.authApp = new PublicClientApplication(msalConfig);

    this.authApp
        .handleRedirectPromise()
    .then(resultHandler);


    // this function will be called when the resultHandler doesn't return a value
    // and will check the state
    // actually this is just necessary in case when the user is logged in and refreshes the page
    function checkAndSetInitialState() {
        if (isInIframe()) {
            return;
        }

        let initialState = false;

        const accounts = this.authApp.getAllAccounts();

        if (accounts.length === 1) {
            const ssoRequest = {
                scopes: [],
                loginHint: accounts[0].username,
            } as AuthorizationUrlRequest;

            try {
                const authResult = await this.authApp.ssoSilent(ssoRequest);
                initialState = !!authResult;
            } catch (error) {
                // ----------------------------
                // HERE IS THE BUG IN IE11
                // ----------------------------
                // other expected causes: too many requests, network error or so
            }
        }

        this.isAuth$.next(initialState);
    }

The logger also doesn't produce any output:

            system: {
                logger: new Logger({
                    loggerCallback: this.msalLogger,
                    piiLoggingEnabled: false,
                    logLevel: LogLevel.Verbose,
                }),
                ...

I also tried with loggerOptions: { ... }

**Comment by jmckennon** [comment]
@daniel-seitz If you're using redirect flows, you should not call `ssoSilent` until `handleRedirectPromise` is completed. We have an example of this in [our sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/d1b306f8acf438d62033b045520e30f81f5bcd55/samples/msal-browser-samples/VanillaJSTestApp2.0/app/ssoSilent/auth.js#L29). This is probably causing the error you're seeing.

> I saw that the retrieved access tokens have the ms graph ressource uri as aud claim and not the applications clientId, is this also a bug?

Since you are not passing any scopes, it's expected that the audience is set to Microsoft Graph. This is intended behavior.

**Comment by daniel-seitz** [comment]
Thank you for the answers @jmckennon,

ssoSilent throws this error (monitor_window_timeout) even if I call it after handleRedirectPromise is finished.
I also found this out and adapted the code. Please see the code of the more _recent comment_ instead.

I tried already to switch off protected mode of my IE11.
I also tried with acquireTokenSilent as outlined in the example that you provided.

**Comment by jmckennon** [comment]
@daniel-seitz Can you take try taking a look at network traces/debugging to see where the bottleneck is? I've still been unable to repro this, unfortunately. You can send a fiddler trace to the email on my profile, as well.

Did you also try to circumvent this by extending the iframe timeout after all these changes? Might be worth trying again.

**Comment by daniel-seitz** [comment]
@jmckennon I debugged it and found the the cause.

In SilentHandler lines 157 and 158 you set:

    if (window.frames && window.frames[iframeId]) {
        authFrame = window.frames[iframeId];
    }

This overwrites what _"authFrame = document.getElementsByTagName("body")[0].appendChild(ifr);"_ set a few lines above. Which seems to work fine - usually.

The IE11 here however doesn't retrieve the wanted Iframe but instead the window object and hence nothing else works afterwards (the src attribute of the window object ist being set and not of the iframe...).

If I uncomment it works.

Can you clarify what the intention for these were? @pkanher617 might know. Getting the frame via window.frames[iframeId] seems odd according to the documentation I found (where they use an index), i.e. https://developer.mozilla.org/en-US/docs/Web/API/Window/frames.

I assume these lines should be moved up into the else if block after line 154.
An else block is missing here as well. 

Please let me know if you need any support here! And thank you for your efforts.

**Comment by jasonnutter** [comment]
@daniel-seitz Thanks for that information. This should be resolved in #2290, as those lines will be removed: https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/2290/files#diff-cf2c9a08cdff88ae830746eba108729cL157-L159

**Comment by daniel-seitz** [comment]
@jmckennon @jasonnutter thank you! 

I tested and it works now. (Even without changing the IE security settings.)

One last question since I have found different examples but no documentation about it:

- Since the recent version works for IE11, is **calling aquireTokenSilent after ssoSilent** necessary anyway, or.. when should I do this? ssoSilent already returns Promise<AuthenticationResult>, doesn't _aquireTokenSilent_ do the same thing anyway?


Here https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/d1b306f8acf438d62033b045520e30f81f5bcd55/samples/msal-browser-samples/VanillaJSTestApp2.0/app/ssoSilent/auth.js
you call **acquireTokenSilent** after ssoSilent. It states that this is just necessary in order to support IE.

    ...
    myMSALObj.ssoSilent(silentRequest).then(() => {
            const currentAccounts = myMSALObj.getAllAccounts();
            accountId = currentAccounts[0].homeAccountId;
            showWelcomeMessage(currentAccounts[0]);

            // call aquireTokenSilent
            getTokenRedirect(loginRequest, currentAccounts[0]);  
        }).catch(error => {
    ....

And here https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-browser-samples/TypescriptTestApp2.0/src/AuthModule.ts you don't use it. 

    this.myMSALObj.ssoSilent(this.silentLoginRequest).then(() => {
            this.account = this.getAccount();
            UIManager.showWelcomeMessage(this.account);
    }).catch(error => {
